### PR TITLE
Query for a given recipe ID

### PR DESCRIPTION
**Pull request by fredex42**
## What does this change?

Add a backend endpoint to get the most recent version of a given recipe ID and the dynamo index to support it

## How to test

- Deploy to CODE
- Make sure you have an API key. You can do this by going to API Gateway in the CAPI account; then `Usage Plans`, `recipes-backend-CODE`, `Associated API Keys`, `Add API Key` (orange button on the right).
- Find a valid recipe UID by running `curl https://recipes.code.dev-guardianapis.com/index.json | jq .recipes`
- Query it by running: 
```
curl https://recipes.code.dev-guardianapis.com/api/content/by-uid/e0a15d7467dde905965da7c52489efe009727422 --header 'X-Api-Key: {api-key}' -D-
```

You should see something like this:
```
HTTP/2 302 
content-type: text/plain; charset=utf-8
x-amzn-requestid: d0a35dec-b7cc-4527-ace3-065e96fa9288
x-amzn-remapped-content-length: 74
x-amz-apigw-id: UN7iCHDWjoEEnwA=
location: /content/z_Xbvog-rK4j2fAEXZOi1O4GeGTmo22my7SLtlWMyT4
x-powered-by: Express
x-amzn-trace-id: Root=1-65e8a60c-193fb96c01512f876c386ed5;Parent=7341af7482025adb;Sampled=0;lineage=a74169f2:0
accept-ranges: bytes
date: Wed, 06 Mar 2024 17:21:16 GMT
via: 1.1 varnish
x-served-by: cache-lcy-eglc8600050-LCY
x-cache: MISS
x-cache-hits: 0
x-timer: S1709745677.659857,VS0,VE315
vary: Accept
strict-transport-security: max-age=300
alt-svc: h3=":443";ma=86400,h3-29=":443";ma=86400,h3-27=":443";ma=86400
content-length: 74

Found. Redirecting to /content/z_Xbvog-rK4j2fAEXZOi1O4GeGTmo22my7SLtlWMyT4
```

To follow the redirect, re-run with `-L`:
```
curl https://recipes.code.dev-guardianapis.com/api/content/by-uid/e0a15d7467dde905965da7c52489efe009727422 --header 'X-Api-Key: {api-key}' -L
```

You should see the JSON of the recipe

## How can we measure success?

Easier to debug problematic recipes

## Have we considered potential risks?

n/a

### CODE REVIEW

File: lambda/rest-endpoints/src/app.ts [hunk 1]
```diff
@@ -65,4 +66,37 @@ router.post('/api/curation/:region/:variant', (req: Request<CurationParams>, res
...
+    validateComposerParams(req.params);
+  } catch(e) {
+    console.log("Provided params ", req.params, " did not validate");
+    resp.status(400).json({status: "error", detail: "invalid recipe immutable ID"});
+    return;
+  }
+
+  recipeByUID(req.params.immutableId).then(result=>{
+    if(result) {
+      console.log(`Debug: Query for ${req.params.immutableId} returned ${JSON.stringify(result)}`);
```
Comment (emdash-ie) [inline lambda/rest-endpoints/src/app.ts:92]: Does using `console.debug` work here, and is the behaviour any different?

**Review by emdash-ie — APPROVED** [review]

**Review by jonathonherbert — APPROVED** [review]
Tested in CODE, works as expected, neat 👍

File: lambda/rest-endpoints/src/app.ts (hunk 1 above)
Comment (fredex42) [reply lambda/rest-endpoints/src/app.ts:92]: I'll drop the line. `console.debug` does work, although if I remember correclty the behaviour is exactly the same.

File: lambda/rest-endpoints/src/app.ts [hunk 2]
```diff
@@ -65,4 +66,37 @@ router.post('/api/curation/:region/:variant', (req: Request<CurationParams>, res
...
+
+function validateComposerParams(params: RecipeIdParams) {
+  const checker = /^[\w\d]+$/;
+
+  if(!params.immutableId.match(checker)) {
+    throw new Error("Invalid composer ID");
+  }
+}
+
+router.get('/api/content/by-uid/:immutableId', (req: Request<RecipeIdParams>, resp) => {
```
[the remark this answers is not in the document]
Comment (fredex42) [reply lambda/rest-endpoints/src/app.ts:81]: fair enough

**Pull request merged**